### DISCUSSION TITLE
Why does brew put things in bash_completions.d

### DISCUSSION THREAD

**Discussion by evanpurkhiser** [Homebrew/discussions#4202] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 3.6.21
ORIGIN: https://github.com/Homebrew/brew
HEAD: 932d2cf3b77c9439a57b6a43577fc8d3b6399a62
Last commit: 6 days ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 738de738346a8ae94f4b03abcbeb3e409d4b9d34
Core tap last commit: 2 days ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: nvim
HOMEBREW_MAKE_JOBS: 8
HOMEBREW_NO_AUTO_UPDATE: set
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_blizzard_avalanche
Clang: 14.0.0 build 1400
Git: 2.39.1 => /opt/homebrew/bin/git
Curl: 7.85.0 => /usr/bin/curl
macOS: 13.1-arm64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

I have been trying to improve my shells startup time. To this end, I have been investigating bash completion, since that was the absolute slowest thing blocking my prompt from appearing immediately.

I've replaced [`bash-completion@1`](https://formulae.brew.sh/formula/bash-completion) with the "new" [`bash-completion@2`](https://formulae.brew.sh/formula/bash-completion@2), which greatly improves startup time by lazy loading completions instead of attempting to load ALL of them at startup.

I noticed it was still eagar loading some completions.. So I disabled eager loading of completions from the "compatibility directory"

```sh
export BASH_COMPLETION_COMPAT_DIR=/dev/null
```

This made my shell start basically instantly. Yay!

However...

From bash-completion@2's docs:

> The other directory which is only present for backwards compatibility, its usage is no longer recommended, is compatdir (get it with pkg-config --variable=compatdir bash-completion). From this directory, files are loaded eagerly when bash_completion is loaded.

With homebrew this is [`$HOMEBREW_PREFIX/etc/bash_completion.d`](https://github.com/Homebrew/homebrew-core/blob/65081b0710f4bf5c4e10bd34bf68b57ed08356ef/Formula/bash-completion%402.rb#L41-L42).

**This is also where homebrew configures pretty much all packages to put their bash completions**. By disabling this I lose almost all of the completions for other software I have installed through brew!

### Some questions

1. Why is that? Why is homebrew using this "legacy compatibility" directory? I do not want these to eager load, it is slow. Why is homebrew not packaging completion files into `$HOMEBREW_PREFIX/share/bash-completion/completions` like other unix distros?

2. I have `git` installed via homebrew. **Why is `git-prompt.sh` in this folder?** It doesn't look anything like a completion script...

3. Why do git completions get installed as `git-completions.bash` and not `git.bash`?

### Workarounds

Right now I manually link these scripts into the user `bash-completion/completions` folder.

See: https://github.com/evanpurkhiser/dots-personal/commit/cba01c381db20bd1fa50ae22caf620996875e7f5

Hoping to get some clarity of why things are the way they are, I couldn't find much other discussion other than this thread https://github.com/Homebrew/homebrew-core/issues/36377

**Comment by SMillerDev** [comment]
https://docs.brew.sh/Shell-Completion

It's the documented approach

**Comment by mbentley** [comment]
So one thing that I do to allow for dynamic loading of completions is by this:
1. Setting `export BASH_COMPLETION_COMPAT_DIR="/dev/null"` as you did; this will break loading of any completions
2. Create a directory `${HOME}/.local/share/bash-completion`.  
3.  Symlink `$(brew --prefix)/etc/bash_completion.d` as `completions` in `${HOME}/.local/share/bash-completion`.  This is where [`BASH_COMPLETION_USER_DIR`](https://github.com/scop/bash-completion/blob/827c3d193d598d8e0ea15c26f235f0ab0732b121/README.md?plain=1#L129) defaults to.
   * So now, you have your completions in `${HOME}/.local/share/bash-completion/completions`
4. As of bash-completion@2 2.12, I now also need to have my .bashrc source `[[ -r "$(brew --prefix)/etc/bash_completion.d/000_bash_completion_compat.bash" ]] && . "$(brew --prefix)/etc/bash_completion.d/000_bash_completion_compat.bash"`
   * This is because what I am doing is really a hack to get loading on demand.


So to summarize, in my `.bashrc`, I have:

```bash
# manually set brew prefix for speed
_BREW_PREFIX="/usr/local"

# override this to null
export BASH_COMPLETION_COMPAT_DIR="/dev/null"

# source bash completion
[[ -r "${_BREW_PREFIX}/etc/profile.d/bash_completion.sh" ]] && . "${_BREW_PREFIX}/etc/profile.d/bash_completion.sh"

# as of bash-completion 2.12, we need to manually source this one file
[[ -r "${_BREW_PREFIX}/etc/bash_completion.d/000_bash_completion_compat.bash" ]] && . "${_BREW_PREFIX}/etc/bash_completion.d/000_bash_completion_compat.bash"
```


And just to show the symlink setup:

```bash
$ ls -l ${HOME}/.local/share/bash-completion
total 0
lrwxr-xr-x@ 1 mbentley  staff  32 Mar 23  2023 completions@ -> /usr/local/etc/bash_completion.d

$ ls -l ${HOME}/.local/share/bash-completion/completions/
total 304
lrwxr-xr-x  1 mbentley  admin     91 Feb 29 17:36 000_bash_completion_compat.bash@ -> ../../Cellar/bash-completion@2/2.12.0/etc/bash_completion.d/000_bash_completion_compat.bash
lrwxr-xr-x@ 1 mbentley  admin     18 Feb 29 19:02 aws@ -> aws_bash_completer
lrwxr-xr-x  1 mbentley  admin     68 Feb 29 17:33 aws_bash_completer@ -> ../../Cellar/awscli/2.15.24/etc/bash_completion.d/aws_bash_completer
lrwxr-xr-x@ 1 mbentley  admin     54 Feb  7 10:29 az@ -> ../../Cellar/azure-cli/2.57.0/etc/bash_completion.d/az
lrwxr-xr-x@ 1 mbentley  admin     36 Mar 22  2023 brew@ -> ../../Homebrew/completions/bash/brew
lrwxr-xr-x@ 1 mbentley  admin     70 Mar 31  2023 docker@ -> /Applications/Docker.app/Contents/Resources/etc/docker.bash-completion
lrwxr-xr-x@ 1 mbentley  admin     47 Feb 20 07:05 gh@ -> ../../Cellar/gh/2.44.1/etc/bash_completion.d/gh
lrwxr-xr-x@ 1 mbentley  admin     19 Mar 23  2023 git@ -> git-completion.bash
lrwxr-xr-x@ 1 mbentley  admin     65 Feb 26 09:13 git-completion.bash@ -> ../../Cellar/git/2.44.0/etc/bash_completion.d/git-completion.bash
lrwxr-xr-x@ 1 mbentley  admin     59 Feb 26 09:13 git-prompt.sh@ -> ../../Cellar/git/2.44.0/etc/bash_completion.d/git-prompt.sh
lrwxr-xr-x@ 1 mbentley  admin     51 Feb 23 09:32 helm@ -> ../../Cellar/helm/3.14.2/etc/bash_completion.d/helm
lrwxr-xr-x@ 1 mbentley  admin     47 Feb 20 07:07 jc@ -> ../../Cellar/jc/1.25.1/etc/bash_completion.d/jc
lrwxr-xr-x@ 1 mbentley  admin     49 Feb 20 07:07 k9s@ -> ../../Cellar/k9s/0.31.9/etc/bash_completion.d/k9s
lrwxr-xr-x  1 mbentley  admin     57 Jan 19 09:38 kompose@ -> ../../Cellar/kompose/1.32.0/etc/bash_completion.d/kompose
lrwxr-xr-x@ 1 mbentley  admin     52 Feb 26 09:12 npm@ -> ../../Cellar/node/21.6.2_1/etc/bash_completion.d/npm
lrwxr-xr-x@ 1 mbentley  admin     61 Feb 28 15:57 parallel@ -> ../../Cellar/parallel/20240222/etc/bash_completion.d/parallel
lrwxr-xr-x@ 1 mbentley  admin     61 Nov 25 05:11 pygmentize@ -> ../../Cellar/pygments/2.17.2/etc/bash_completion.d/pygmentize
lrwxr-xr-x@ 1 mbentley  admin     56 Feb 12 09:25 tanzu@ -> ../../Cellar/tanzu-cli/1.2.0/etc/bash_completion.d/tanzu
lrwxr-xr-x@ 1 mbentley  admin     52 Feb 20 07:05 tilt@ -> ../../Cellar/tilt/0.33.11/etc/bash_completion.d/tilt
lrwxr-xr-x@ 1 mbentley  admin     48 Feb 20 07:07 tmux@ -> ../../Cellar/tmux/3.4/etc/bash_completion.d/tmux
lrwxr-xr-x@ 1 mbentley  admin     47 Feb 26 09:10 yq@ -> ../../Cellar/yq/4.42.1/etc/bash_completion.d/yq
lrwxr-xr-x@ 1 mbentley  admin     50 Feb  7 10:32 ytt@ -> ../../Cellar/ytt/v0.48.0/etc/bash_completion.d/ytt
```

You'll notice there are a few completion files that do not follow the correct format for dynamically loading (they have to be exactly matching the binary name) so things like `aws` and `git` I have to create a symlink to the symlink so it loads.
